server/src/main/scala/io/delta/sharing/server/model.scala

```diff
@@ -135,3 +135,130 @@ object Action {
   // Basically delta sharing doesn't support write for now.
   val maxWriterVersion = 0
 }
+
+/**
+ * Actions defined in delta format, used in response when requested format is delta.
+ */
+
+sealed trait DeltaAction {
+  /** Turn this object to the [[DeltaSingleAction]] wrap object. */
+  def wrap: DeltaSingleAction
+}
+
+/**
+ * Used to block older clients from reading the shared table when backwards
+ * incompatible changes are made to the protocol. Readers and writers are
+ * responsible for checking that they meet the minimum versions before performing
+ * any other operations.
+ *
+ * Since this action allows us to explicitly block older clients in the case of a
+ * breaking change to the protocol, clients should be tolerant of messages and
+ * fields that they do not understand.
+ */
+case class DeltaProtocol(minReaderVersion: Int) extends DeltaAction {
+  override def wrap: DeltaSingleAction = DeltaSingleAction(protocol = this)
+}
+
+/**
+ * DeltaAddFile used in delta sharing protocol, copied from AddFile in delta.
+ *   Adding 4 delta sharing related fields: id/version/timestamp/expirationTimestamp,
+ *       which will be redacted at the client side before writing to the delta log.
+ *   Ignoring 1 field: tags.
+ */
+case class DeltaAddFile(
+    path: String,
+    id: String,
+    @JsonInclude(JsonInclude.Include.ALWAYS)
+    partitionValues: Map[String, String],
+    size: Long,
+    modificationTime: Long,
+    dataChange: Boolean,
+    @JsonRawValue
+    stats: String = null,
+    version: java.lang.Long = null,
+    timestamp: java.lang.Long = null,
+    expirationTimestamp: Long) extends DeltaAction {
+  require(path.nonEmpty)
+
+  override def wrap: DeltaSingleAction = DeltaSingleAction(add = this)
+}
+
+/**
+ * DeltaRemoveFile used in delta sharing protocol, copied from RemoveFile in delta.
+ *   Adding 4 delta sharing related fields: id/version/timestamp/expirationTimestamp,
+ *       which will be redacted at the client side before writing to the delta log.
+ *   Ignoring 1 field: tags.
+ */
+case class DeltaRemoveFile(
+    path: String,
+    id: String,
+    deletionTimestamp: Option[Long],
+    dataChange: Boolean = true,
+    extendedFileMetadata: Boolean = false,
+    partitionValues: Map[String, String] = null,
+    size: Option[Long] = None,
+    version: Long,
+    timestamp: Long,
+    expirationTimestamp: Long) extends DeltaAction {
+  override def wrap: DeltaSingleAction = DeltaSingleAction(remove = this)
+}
+
+/**
+ * DeltaAddCDCFile used in delta sharing protocol, copied from AddCDCFile in delta.
+ *   Adding 4 delta sharing related fields: id/version/timestamp/expirationTimestamp,
+ *       which will be redacted at the client side before writing to the delta log.
```

Comment:
Consider expanding on why we are sending these?

This comment assumes a certain client implementation (the one we are working on currently).
We should perhaps generalize this comment:
   "If the client uses delta kernel, it should redact these fields as needed".

Reply:
good point, it assumes client side operatoins, but I think these will be part of the protocol, as client need these fields to construct the delta log correctly.